tests/db_functions/tests.py

```diff
@@ -99,6 +101,70 @@ def test_coalesce_ordering(self):
             lambda a: a.name
         )
 
+    def test_greatest(self):
+        now = timezone.now()
+        before = now - timedelta(hours=1)
+
+        Article.objects.create(
+            title="Testing with Django",
+            written=before,
+            published=now,
+        )
+
+        articles = Article.objects.annotate(
+            last_updated=Greatest('written', 'published'),
+        )
+        self.assertEqual(articles.first().last_updated, now)
+
+    @skipUnlessDBFeature('greatest_least_ignores_nulls')
+    def test_greatest_ignores_null(self):
+        now = timezone.now()
+
+        Article.objects.create(title="Testing with Django", written=now)
+
+        articles = Article.objects.annotate(
+            last_updated=Greatest('written', 'published'),
+        )
+        self.assertEqual(articles.first().last_updated, now)
+
+    @skipIfDBFeature('greatest_least_ignores_nulls')
+    def test_greatest_propogates_null(self):
+        now = timezone.now()
+
+        Article.objects.create(title="Testing with Django", written=now)
+
+        articles = Article.objects.annotate(
+            last_updated=Greatest('written', 'published'),
+        )
+        self.assertIsNone(articles.first().last_updated)
+
+    def test_greatest_all_null(self):
+        Article.objects.create(title="Testing with Django", written=timezone.now())
+
+        articles = Article.objects.annotate(last_updated=Greatest('published', 'updated'))
+        self.assertIsNone(articles.first().last_updated)
+
+    def test_greatest_one_expressions(self):
+        with self.assertRaisesMessage(ValueError, 'Greatest must take at least two expressions'):
+            Greatest('written')
+
+    def test_greatest_related_field(self):
+        author = Author.objects.create(name='John Smith', age=45)
+        Fan.objects.create(name='Margaret', age=50, author=author)
+
+        authors = Author.objects.annotate(
+            highest_age=Greatest('age', 'fans__age'),
+        )
+        self.assertEqual(authors.first().highest_age, 50)
+
+    def test_greatest_update(self):
+        author = Author.objects.create(name='John Smith', age=45)
+        Fan.objects.create(name='Margaret', age=50, author=author)
+
+        Author.objects.update(age=Greatest('age', 'fans__age'))
+        author.refresh_from_db()
+        self.assertEqual(author.age, 50)
```

Comment:
@jarshwah Is this the sort of test you were thinking of?
It currently fails with a `FieldError` that looks related to [#14104](https://code.djangoproject.com/ticket/14104).

```python-traceback
Traceback (most recent call last):
  File "/home/ian/dev/django/tests/db_functions/tests.py", line 164, in test_greatest_update
    Author.objects.update(age=Greatest('age', 'fans__age'))
  File "/home/ian/dev/django/django/db/models/manager.py", line 125, in manager_method
    return getattr(self.get_queryset(), name)(*args, **kwargs)
  File "/home/ian/dev/django/django/db/models/query.py", line 619, in update
    rows = query.get_compiler(self.db).execute_sql(CURSOR)
  File "/home/ian/dev/django/django/db/models/sql/compiler.py", line 1065, in execute_sql
    cursor = super(SQLUpdateCompiler, self).execute_sql(result_type)
  File "/home/ian/dev/django/django/db/models/sql/compiler.py", line 832, in execute_sql
    sql, params = self.as_sql()
  File "/home/ian/dev/django/django/db/models/sql/compiler.py", line 1018, in as_sql
    val = val.resolve_expression(self.query, allow_joins=False, for_save=True)
  File "/home/ian/dev/django/django/db/models/expressions.py", line 491, in resolve_expression
    c.source_expressions[pos] = arg.resolve_expression(query, allow_joins, reuse, summarize, for_save)
  File "/home/ian/dev/django/django/db/models/expressions.py", line 448, in resolve_expression
    return query.resolve_ref(self.name, allow_joins, reuse, summarize)
  File "/home/ian/dev/django/django/db/models/sql/query.py", line 1416, in resolve_ref
    raise FieldError("Joined field references are not permitted in this query")
django.core.exceptions.FieldError: Joined field references are not permitted in this query
```